### PR TITLE
Fix visual glitch when hunter is waiting for the animal

### LINKED ISSUE #1126: Assertion failure in CalcRelative
this hasn't happend for such a long time, that I wasn't running the game through gdb. I'm going to reload the last save game now, and have the replay attached. 

[2019-10-30_20-59-24.rpl.zip](https://github.com/Return-To-The-Roots/s25client/files/3790923/2019-10-30_20-59-24.rpl.zip)

### PR DESCRIPTION
**Pull request by Flamefire**
When the hunter arrived at the shooting pos but the animal was not yet
ready, the hunter was still drawn as "walking" making it walk past its
destination point (due to interpolation using the "wait" event) and
warped back once the animal was ready.
This also triggered an assertion checking for walking draw while not
walking

Fixes #1126

### CODE REVIEW

File: libs/s25main/figures/nofHunter.cpp [hunk 1]
```diff
@@ -60,6 +60,9 @@ nofHunter::nofHunter(SerializedGameData& sgd, const unsigned obj_id) : nofBuildi
         animal = sgd.PopObject<noAnimal>(GOT_ANIMAL);
         shootingPos = sgd.PopMapPoint();
         shooting_dir = Direction::fromInt(sgd.PopUnsignedChar());
+        // https://github.com/Return-To-The-Roots/s25client/issues/1126
+        if(sgd.GetGameDataVersion() < 4 && state == STATE_HUNTER_FINDINGSHOOTINGPOINT && pos == shootingPos)
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:64]: How would you name it? I can't find a meaningful name and using `VERSION_4` doesn't really help.

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:64]: GAMEDATA_VERSION4 - and make a comment whats not in there below that point?

File: libs/s25main/figures/nofHunter.cpp [hunk 2]
```diff
@@ -463,34 +472,25 @@ void nofHunter::AnimalLost()
...
-            // nach Haue laufen
-            StartWalkingHome();
-        }
-        break;
+        case STATE_HUNTER_WALKINGTOCADAVER: StartWalkingHome(); break;
         case STATE_HUNTER_SHOOTING:
         case STATE_HUNTER_EVISCERATING:
-        {
+        case STATE_HUNTER_WAITING_FOR_ANIMAL_READY:
             // Arbeits-Event abmelden
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:475]: I like it better this way:
- Short blocks (single calls) get automatically put onto 1 line
- indentation is still preserved
- `break` is on the same level as the code and no thinking whether to put it inside or outside the braces.

In total: As we use clang-format the braces don't serve any purpose anymore except when they are used to introduce a scope (for variables)

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:475]: yes then simply put those multiple commands into a method und call it "single lined" - makes it more readable

File: libs/s25main/figures/nofHunter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:64]: I don't see the advantage of `sgd.GetGameDataVersion() < GAMEDATA_VERSION4` over the current version. Can you explain a bit more what you want to achieve?

If it is documenting what changed then we could do that where the current version is defined as a simple list without introducing new constants

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:64]: but that way you can search for it and find information about it. with a simple "4" you don't know anything.

File: libs/s25main/figures/nofHunter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:475]: Why is it more readable? Now you see that you get 2 similar behaviors in the handling of `AnimalLost`. It is nicely indented so it cannot be confused with a different label and 3 statements are short. Better than making up a single-use function named `StopEventAndStartWalkingHome` or so.

File: libs/s25main/figures/nofHunter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:64]: This is kinda intentional: An increase in the GameDataVersion means that some saving code has changed and loading code in (usually) 1 place needs to be adapted to load earlier versions. So this piece of code guarded by `sgd.GetGameDataVersion() < 4` is all the there is to the "4". Other code doesn't need to know that "4" even exists and there is nothing more than this 2 lines of code to say about that 4.

Introducing the constant(s) means that all cpp files using the `sgd` would need to be recompiled even so it doesn't matter to them.

So I'd rather keep this as some information that is local to the code using it.

It is simply a version hence it is a simple number. Same as other code using things like `#if LIBFOO_VERSION < 0304` for compatibility with older library versions does not introduce constants like `VERSION_0304` with comments explaining what the code using that constant is doing.

Hope that clears it up

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:64]: but normally you have a list at the place where libfoo_version is defined, where its stated what has been changed. Thats what I want to have.

So:
- At the place where GameDataVersion is defined: a VERSION4 = 4 constant with a comment like "added ... new hunter state"
- and then use that constant there on the < 4. This way you can search for VERSION4 and find the code pieces. if you can only search for "4" you willl get a lot of matches. thats what I dont want

[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:64]: Rarely. Maybe there is a changelog at the website of the lib but that might not mention why you check for a specific version at that place. 

There are 2 points: adding the constant affects many files which I don't want. And second you don't need to search for "4". All information is right here: on older versions you need to fix the state. Full description on why is in the linked issue.

But OK as a compromise I can add a changelog at the point where the current version is defined and add a comment to GetGameDataVersion() with a reference to that. This matches the libfoo_version handling above and does not cause needless recompiles

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:64]: ok!

File: libs/s25main/figures/nofHunter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:475]: I though I've seen the same few lines multiple times there perhaps its good to create a method for it? - i.e "StartWalkingHome" - and only stop event if set, or so?

File: libs/s25main/figures/nofHunter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:64]: BTW: Would it make sense to split the serializer into a serializer and deserializer (same for SerializedGameData/GameDataSerializer)? It bugs me that there are comments like `(-> only valid during writing)` and both Push/Pop functions while you are only supposed to use one of them but this isn't enforced...

[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:65]: https://github.com/Return-To-The-Roots/s25client/pull/1268/commits/09cdbfb10678d38d9e49412df775d8df2f2e26e8

File: libs/s25main/figures/nofHunter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/s25main/figures/nofHunter.cpp:451]: I merged StartWalkingHome and WalkingHome with a renaming to match other `HandleState*` functions which reduces the code to 2 lines. Stopping the event in the function would make sense but we usually just overwrite `current_ev` with the new event and removing the currently handled event is disallowed. But that is what would happen. Example: EventGatherFood triggers the handler, hunter picks up food, StartWalkingHome is called, would try to remove EventGatherFood

As there is only 1 place where the current event should be removed I left it as-is to make it explicit.

[the remark this answers is not in the document]
Comment (Flow86) [reply libs/s25main/figures/nofHunter.cpp:451]: ok

**Comment by Flamefire** [comment]
Checks passed. @Flow86 Can you merge please?

**Pull request merged**